record_imu_cam/plugin.cpp

```diff
@@ -0,0 +1,98 @@
+#include "common/plugin.hpp"
+#include "common/switchboard.hpp"
+#include "common/data_format.hpp"
+
+#include <fstream>
+#include <boost/filesystem.hpp>
+#include <iomanip>
+#include <opencv2/core/mat.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/imgproc.hpp>
+using namespace ILLIXR;
+
+class record_imu_cam : public plugin {
+public:
+	record_imu_cam(std::string name_, phonebook* pb_)
+		: plugin{name_, pb_}
+		, sb{pb->lookup_impl<switchboard>()}
+		, record_data{get_record_data_path()}
+		, cam0_data_dir{record_data / "cam0" / "data"}
+		, cam1_data_dir{record_data / "cam1" / "data"}
+	{
+		// check folder exist, if exist delete it
+		boost::filesystem::remove_all(record_data);
+
+		// create imu0 directory
+		boost::filesystem::path imu_dir = record_data / "imu0";
+		boost::filesystem::create_directories(imu_dir);
+		std::string imu_file = imu_dir.string() + "/data.csv";
+		imu_wt_file.open(imu_file, std::ofstream::out);
+		imu_wt_file << "#timestamp [ns],w_x [rad s^-1],w_y [rad s^-1],w_z [rad s^-1],a_x [m s^-2],a_y [m s^-2],a_z [m s^-2]" << std::endl;
+
+		// create cam0 directory
+		boost::filesystem::create_directories(cam0_data_dir);
+		std::string cam0_file = (record_data / "cam0" / "data.csv").string();
+		cam0_wt_file.open(cam0_file, std::ofstream::out);
+		cam0_wt_file << "#timestamp [ns],filename" << std::endl;
+
+		// create cam1 directory
+		boost::filesystem::create_directories(cam1_data_dir);
+		std::string cam1_file = (record_data / "cam1" / "data.csv").string();
+		cam1_wt_file.open(cam1_file, std::ofstream::out);
+		cam1_wt_file << "#timestamp [ns],filename" << std::endl;
+
+		sb->schedule<imu_cam_type>(id, "imu_cam", [this](switchboard::ptr<const imu_cam_type> datum, std::size_t){
+			this->dump_data(datum);
+		});
+	}
+
+	void dump_data(switchboard::ptr<const imu_cam_type> datum) {
+		ullong timestamp  = datum->dataset_time;  // long timestamp = datum->time.time_since_epoch().count(); or long timestamp = _m_clock->now().time_since_epoch().count(); 
```

Comment:
We no longer have `dataset_time` in the `imu_cam_type`. I am thinking of changes as the comment, using either `time` ~~or `_m_clock->now()`~~ as the time. Given how we record `time` [here](https://github.com/ILLIXR/ILLIXR/blob/2981204b9b64b6d093ac23ffc8e88aca3e5fc14a/offline_imu_cam/plugin.cpp#L95), it should be able to reserve the relative timing relationship between data points. But we can't get the real dataset timestamp because we lose the "first" (`dataset_first`) time information.

Reply:
Yes, we should use `time` here. `record_imu_cam` is meant for recording the IMU and camera data on a real run, not using `offline_imu_cam`. As such, the `dataset_time` should be equal to the `time` (no adjustment needed). Saving the `time` should be fine.

Reply:
updated